src/Evolution/Systems/ScalarWave/Equations.hpp

```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
                     const tnsr::i<DataVector, Dim, Frame::Inertial>&
                         interface_unit_normal) noexcept;
 };
+
+/*!
+ * \brief Computes the upwind flux
+ *
+ * The upwind flux is given by:
+ * \f{align}
+ * F^*(\Psi) =& 0 \\
+ * F^*(\Pi) =& \frac{1}{2}\left(F(\Pi)_{\mathrm{int}} + F(\Pi)_{\mathrm{ext}}
+ *                    + \Pi_{\mathrm{int}} - \Pi_{\mathrm{ext}}\right) \\
+ * F^*(\Phi_i) =& \frac{1}{2} \left(F(\Phi_i)_{\mathrm{int}}
+ *                   + F(\Phi_i)_{\mathrm{ext}}
+ *                   + (n_i)_{\mathrm{int}} F(\Pi)_{\mathrm{int}}
+ *                   - (n_i)_{\mathrm{ext}} F(\Pi)_{\mathrm{ext}}\right)
+ * \f}
+ * where $F$ is the flux computed in ScalarWave::ComputeNormalDotFluxes
+ */
+template <size_t Dim>
+struct UpwindFlux {
+ private:
+  struct NormalTimesFluxPi {
+    using type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+    static constexpr const char* const label = "UpwindFlux";
+  };
+
+ public:
+  using options = tmpl::list<>;
+  static constexpr OptionString help = {"Class help text"};
+
+  // clang-tidy: non-const reference
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
+
+  using return_tags = tmpl::list<Tags::NormalDotNumericalFlux<Pi>,
+                                 Tags::NormalDotNumericalFlux<Phi<Dim>>,
+                                 Tags::NormalDotNumericalFlux<Psi>>;
+
+  using package_tags =
+      tmpl::list<Tags::NormalDotFlux<Pi>, Tags::NormalDotFlux<Phi<Dim>>, Pi,
+                 NormalTimesFluxPi>;
+
+  using slice_tags = tmpl::list<Pi>;
+
+  void package_data(
+      gsl::not_null<Variables<package_tags>*> packaged_data,
+      const Scalar<DataVector>& normal_dot_flux_pi,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_dot_flux_phi,
+      const Scalar<DataVector>& /*normal_dot_flux_psi*/,
+      const Scalar<DataVector>& pi,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+      const noexcept;
+
+  void operator()(
```

Comment:
`// pseudo-interface: used internally by Algorithm infrastructure, not user-level code`